### ISSUE TITLE
Adding Switches

### ISSUE THREAD

**Issue by Deki4** [PyPSA/PyPSA#89]
Add switches as a new component. Each line would have two switches with status as open or closed. An example csv/dataframe would look as follows:

| name | line | connected_bus | disconnected_bus | status | b0b1 |
| --- | --- | --- | --- | --- | --- | 
| s1 | 63 | 13915_LV | 13915_LV_s1 | open | bus0 |
| s2 | 64 | 13911_LV | 13911_LV_s2 | open | bus1 |

- Two open switches place a line out of service by removing from the net.lines df and into an 'os_lines' dataframe of the network object for accounting purposes. This way the line can be appended to the net.lines df from the os_lines df if brought back into service. 
- One open switch creates an auxiliary bus (similar to pandapower) referred to as "disconnected_bus" in the above mentioned dataframe. 

I currently have working code for the above logic for my own "grid" class which I inherited from PyPSA's Network class. Would be great to hear your thoughts.

**Comment by fneum** [comment]
Have a look at PR #77 which introduces a flag stating whether a line is operational or not. #77 should somewhat implement what you are thinking of albeit possibly simpler. Can you tell me why you would go for two switches instead of one?

**Comment by fneum** [comment]
I take it that you would *not* have the switches as variables in the problem but rather use them to determine how to build the problem?

**Comment by Deki4** [comment]
I should also mention that this logic of 'in and out of service' lines is carried out by a switching() method in the network class. By having two switches we allow the capacitance of the line to be accounted for in the power flow. This is because when a single switch is closed (and the other open), the auxiliary bus is created at the same geographical coordinates as the previously connected bus thus preserving the lines length. This also allows disconnection from other parts of the grid and the possibility to connect other/new components to this new auxiliary bus. 
I figured it would be easier to explain with plots of a grid I am using.
These illustrate the case where one switch is open and one closed on the line. 
![grid1](https://user-images.githubusercontent.com/20196262/63110787-80476500-bf8c-11e9-8fea-3f5bc8fca909.png)
![grid2](https://user-images.githubusercontent.com/20196262/63110806-89d0cd00-bf8c-11e9-83fb-d9fc2014db44.png)

**Comment by fneum** [comment]
If you are willing to share the code you already have, I might look into integrating such a function in PyPSA. Best case you prepare a [pull request](https://help.github.com/en/articles/about-pull-requests) with a proposal how this function could be cleanly integrated in PyPSA.

**Comment by jankaeh** [comment]
Hi, I have been working on implementing a more general logic for switches. https://github.com/jankaeh/PyPSA/tree/switches
I added the component "Switch" as a logical element, that does determine the network topology but does not have impedance. The file logic.py defines the functions needed for this.
I added switches to the components-documentation and try to explain the particularities that come along with it: https://github.com/jankaeh/PyPSA/blob/switches/doc/components.rst#switch
I'll add an example to show the usage hopefully tomorrow. I'd appreciate your feedback - which is probably easier to give whith an existing example... One point that's not adressed yet is taking components "out of service", maybe the planned attribute opreational (once implemented) could be updated when opening or closing switches.
